### PR TITLE
feat(simulation): add simulator to the bench

### PR DESCRIPTION
**Pull request by wivwiv**
#### What is the new behavior?

Users can place a custom script in the simulator directory and then use it to generate simulation data in the bench pub with the -sim option.

This helps you to achieve a simulated performance test that is closer to the production.

```bash
mqttx bench pub -sim tesla -t t/1 -c 1

mqttx sub -t t/1
{"car_id":"JK6ZSF1A59HS94294","display_name":"Ines's Tesla","model":"Y","trim_badging":"ullam","exterior_color":"purple","wheel_type":"repudiandae","spoiler_type":"placeat","geofence":"Longview","state":"charging","since":"2023-04-23T18:25:00.558Z","healthy":true,"version":"0.2.9","update_available":false,"update_version":"1.0.2","latitude":"-15.3417","longitude":"127.2175","shift_state":"P","power":-6046,"speed":189,"heading":50,"elevation":3851,"locked":true,"sentry_mode":true,"windows_open":false,"doors_open":true,"trunk_open":true,"frunk_open":false,"is_user_present":false,"is_climate_on":false,"inside_temp":38.4,"outside_temp":44.8,"is_preconditioning":true,"odometer":383942,"est_battery_range_km":653.7,"rated_battery_range_km":916.6,"ideal_battery_range_km":542.9,"battery_level":4,"usable_battery_level":87,"plugged_in":true,"charge_energy_added":74.07,"charge_limit_soc":42,"charge_port_door_open":true,"charger_actual_current":47.44,"charger_power":47,"charger_voltage":240,"charge_current_request":15,"charge_current_request_max":20,"scheduled_charging_start_time":"2023-01-25T02:10:33.559Z","time_to_full_charge":8.42,"tpms_pressure_fl":2.5,"tpms_pressure_fr":3.3,"tpms_pressure_rl":3.2,"tpms_pressure_rr":2.3,"timestamp":1682330389155}
```

#### Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

#### Specific Instructions

Are there any specific instructions or things that should be known prior to review?

### CODE REVIEW

**Comment by Red-Asuka** [comment]
@wivwiv How to introduce an external custom simulator? I see that the directory in the code is obtained using the following method, which is the path of the current file. The simulator needs to be in this directory to work properly. The development environment directory address is `MQTTX/cli/dist/src/simulator`. When users use it, it is packaged into a binary executable file. I am not sure how to reference it specifically.

```ts
const SimulatorFolder = path.join(__dirname, '../simulator')
```

**Comment by Red-Asuka** [comment]
@wivwiv CLI currently does not have a configuration to automatically format code, nor does it use ESLint. It only has a Prettier configuration file. Please use Prettier to format the code before submitting.

File: cli/src/index.ts [hunk 1]
```diff
@@ -355,6 +356,7 @@ export class Commander {
         'the message topic, support %u (username), %c (client id), %i (index) variables',
         parsePubTopic,
       )
+      .option('-sim, --simulator <SIMULATOR>', 'message generation simulator', parseSimulator, '')
```
Comment (Red-Asuka) [inline cli/src/index.ts:359]: Maybe this feature can also be added to the `pub` command

**Comment by ysfscream** [comment]
Thank you for submitting this PR! After reviewing the changes, I have a few suggestions to make the mqttx cli library more flexible and adaptable for different use cases.

1. Remove the 'tesla' data file: As discussed, the 'tesla' file is quite specific to a particular use case. Including it in the library may not be suitable for a general-purpose utility like mqttx cli. I recommend removing this file from the PR.

2. Implement pluggable scenario templates: Instead of including a specific data file, consider introducing a pluggable approach that allows users to create their scenario data templates based on their needs. This would make the library more versatile and cater to a wider range of use cases.

3. Provide examples: To guide users in creating their templates, including a few example templates showcasing different use cases would be helpful. This will assist users in understanding how to use the library effectively and create a more user-friendly experience.

Once these changes are implemented, I believe the PR will be a valuable addition to the mqttx cli library. Thank you for your hard work EMQX PM 😂, and please let me know if you have any questions or concerns!

**Comment by ysfscream** [comment]
1. **Define a template function**: Create a template function that takes the scenario configuration as input and returns an object containing the topic and payload.

```javascript
function defaultTemplate(config) {
    return {
        topic: 'default_topic',
        payload: 'default_payload'
    };
}
```

2. **Allow users to pass a custom template**: Update your `mqttx` CLI tool to accept a custom template function as an optional parameter. If no template is provided, the default implementation will be used.

```javascript
// mqttx-bench.js
const yargs = require('yargs/yargs');
const { hideBin } = require('yargs/helpers');

// Import the defaultTemplate function
const { defaultTemplate } = require('./template');

const argv = yargs(hideBin(process.argv))
    .command('bench pub [config] [template]', 'publish benchmark', (yargs) => {
        yargs
            .positional('config', {
                describe: 'path to the configuration file',
                type: 'string'
            })
            .positional('template', {
                describe: 'path to the custom template function file',
                type: 'string'
            });
    })
    .help()
    .argv;

function runMqttxCli(configPath, templatePath) {
    const config = require(configPath);
    const template = templatePath ? require(templatePath) : defaultTemplate;

    // Use the provided template function to get topic and payload
    const { topic, payload } = template(config);

    // Continue with the rest of the mqttx CLI logic
}
```

3. **Update documentation**: Update your library's documentation to explain how users can create their custom templates as functions and pass the path of the function file as an argument to the `mqttx bench pub` command.

Here's an example of a custom template function implementation in a separate file, let's say `my_custom_template.js`:

```javascript
// my_custom_template.js
function myCustomTemplate(config) {
    return {
        topic: 'my_custom_topic',
        payload: 'my_custom_payload'
    };
}

module.exports = myCustomTemplate;
```

Users can then run the MQTTX CLI tool with their custom template like this:

```bash
mqttx bench pub ./path/to/config.json ./path/to/my_custom_template.js
```

By implementing pluggable scenario templates using functions, you allow users to customize the behavior of your MQTTX CLI tool for different use cases.

**Review by ysfscream — CHANGES_REQUESTED** [review]
Need Update

**Pull request closed without merging**